src/Features/Core/Portable/ExtractMethod/SelectionValidator.cs

```diff
@@ -4,208 +4,201 @@
+            if (controlFlowAnalysisData.EntryPoints.Any())
             {
-                continue;
+                return false;
             }
 
-            // quick exit check
-            // - passed candidate statements
-            if (textSpan.End < stmt.SpanStart)
+            // check something like continue, break, yield break, yield return, and etc
+            if (ContainsNonReturnExitPointsStatements(controlFlowAnalysisData.ExitPoints))
             {
-                break;
+                return false;
             }
 
-            if (stmt.SpanStart <= textSpan.Start)
+            // okay, there is no branch out, check whether next statement can be executed normally
+            var returnStatements = GetOuterReturnStatements(range.Item1.GetCommonRoot(range.Item2), controlFlowAnalysisData.ExitPoints);
+            if (!returnStatements.Any())
             {
-                // keep track spine
-                spine.Add(stmt);
+                if (!controlFlowAnalysisData.EndPointIsReachable)
+                {
+                    // REVIEW: should we just do extract method regardless or show some warning to user?
+                    // in dev10, looks like we went ahead and did the extract method even if selection contains
+                    // unreachable code.
+                }
+
+                return true;
             }
 
-            if (textSpan.End <= stmt.Span.End && spine.Any(s => CanMergeExistingSpineWithCurrent(syntaxFacts, s, stmt)))
-            {
-                // malformed code or selection can make spine to have more than an elements
-                firstStatement = spine.First(s => CanMergeExistingSpineWithCurrent(syntaxFacts, s, stmt));
-                lastStatement = stmt;
+            // okay, only branch was return. make sure we have all return in the selection.
 
-                spine.Clear();
+            // check for special case, if end point is not reachable, we don't care the selection
+            // actually contains all return statements. we just let extract method go through
+            // and work like we did in dev10
+            if (!controlFlowAnalysisData.EndPointIsReachable)
+            {
+                return true;
             }
-        }
 
-        if (firstStatement == null || lastStatement == null)
-        {
-            return null;
+            // there is a return statement, and current position is reachable. let's check whether this is a case where that is okay
+            return IsFinalSpanSemanticallyValidSpan(textSpan, returnStatements, cancellationToken);
         }
 
-        return (firstStatement, lastStatement);
-
-        static bool CanMergeExistingSpineWithCurrent(ISyntaxFacts syntaxFacts, T existing, T current)
-            => syntaxFacts.AreStatementsInSameContainer(existing, current);
-    }
+        protected (TStatementSyntax firstStatement, TStatementSyntax lastStatement)? GetStatementRangeContainingSpan(
```

Comment:
no need for this method to be generic, it was always being passed the TStatementSyntax type.